### PR TITLE
Upgrade rails version

### PR DESCRIPTION
**Pull request by SamJamCul**
Shifting up to Rails 7 - but not yet changing the defaults. Just sorting out breaking features and a few deprecation warnings.

### CODE REVIEW

File: app/queries/reasons_for_rejection_count_query.rb [hunk 1]
```diff
@@ -24,7 +24,7 @@ def reason_counts
     rows = ActiveRecord::Base.connection.exec_query(
       reason_counts_sql,
       'SQL',
-      [[nil, Time.zone.now.beginning_of_month]],
```
Comment (cwrw) [inline app/queries/reasons_for_rejection_count_query.rb:27]: This was removed in https://github.com/rails/rails/commit/2d821ef0c5356d860eba2d2341b4db71b67f6594
From what I can see it preserves the previous behaviour, however worth double checking with @stevehook

File: config/application.rb [hunk 2]
```diff
@@ -89,6 +89,9 @@ class Application < Rails::Application
       'Feature-Policy' => "accelerometer 'none'; ambient-light-sensor: 'none', autoplay: 'none', battery: 'none', camera 'none'; display-capture: 'none', document-domain: 'none', fullscreen: 'none', geolocation 'none'; gyroscope 'none'; magnetometer 'none'; microphone 'none'; midi: 'none', payment 'none'; publickey-credentials-get: 'none', usb 'none', wake-lock: 'none', screen-wake-lock: 'none', web-share: 'none'",
     }
 
+    config.active_record.legacy_connection_handling = false
```
Comment (cwrw) [inline config/application.rb:92]: this is part of the defaults in 6.1, can we do this when we set the defaults on then?

File: config/environments/development.rb [hunk 3]
```diff
@@ -14,9 +14,9 @@
   # Show full error reports.
   config.consider_all_requests_local = true
 
-  # todo
+  # todo not clear what this does!
   # Enable server timing
-  # config.server_timing = true
+  config.server_timing = true
```
Comment (cwrw) [inline config/environments/development.rb:19]: https://github.com/rails/rails/pull/36289

File: config/environments/development.rb [hunk 4]
```diff
@@ -75,16 +75,19 @@
   # Highlight code that triggered database queries in logs.
   config.active_record.verbose_query_logs = true
 
-  # todo
+  # todo not clear what the effect is
   # Suppress logger output for asset requests.
   # config.assets.quiet = true
```
Comment (cwrw) [inline config/environments/development.rb:80]: seems to be introduced in Rails 5, lets remove

File: config/environments/development.rb [hunk 5]
```diff
@@ -75,16 +75,19 @@
...
-  # config.i18n.raise_on_missing_translations = true
+  config.i18n.raise_on_missing_translations = true
 
+  # todo seems handy, puts template names in the html
   # Annotate rendered view with file names.
-  # config.action_view.annotate_rendered_view_with_filenames = true
+  config.action_view.annotate_rendered_view_with_filenames = true
 
+  # todo not clear what this does
   # Uncomment if you wish to allow Action Cable access from any origin.
```
Comment (cwrw) [inline config/environments/development.rb:91]: no need to enable, can remove

File: config/environments/test.rb [hunk 6]
```diff
@@ -44,16 +44,19 @@
...
   # Raise exceptions for disallowed deprecations.
   # config.active_support.disallowed_deprecation = :raise
 
+  # todo no need until we add disallowed deprecations
   # Tell Active Support which deprecation messages to disallow.
   # config.active_support.disallowed_deprecation_warnings = []
 
+  # todo could enable?
   # Raises error for missing translations.
   # config.i18n.raise_on_missing_translations = true
```
Comment (cwrw) [inline config/environments/test.rb:57]: yes lets enable

File: config/environments/test.rb [hunk 7]
```diff
@@ -21,8 +21,8 @@
   # Show full error reports and disable caching.
   config.consider_all_requests_local       = true
 
-  # todo
-  # config.action_controller.perform_caching = false
+  # todo not clear what effect this has
```
Comment (cwrw) [inline config/environments/test.rb:24]: let's remove this one as the default is true

File: config/initializers/permissions_policy.rb [hunk 8]
```diff
@@ -0,0 +1,11 @@
+# Define an application-wide HTTP permissions policy. For further
```
Comment (cwrw) [inline config/initializers/permissions_policy.rb:1]: this one is interesting, could keep to revisit later 🤔

File: config/initializers/inflections.rb [hunk 9]
```diff
@@ -0,0 +1,16 @@
+# Be sure to restart your server when you modify this file.
```
Comment (cwrw) [inline config/initializers/inflections.rb:1]: could move `inflector.rb` setup here

File: config/environments/test.rb (hunk 6 above)
Comment (duncanjbrown) [reply config/environments/test.rb:57]: I think this is on globally in `application.rb`?

File: config/application.rb [hunk 10]
```diff
@@ -32,7 +37,7 @@ class Application < Rails::Application
     config.load_defaults 6.0
     config.autoloader = :zeitwerk
 
-    # Settings in config/environments/* take precedence over those specified here.
+    # Settings in config/environments/* tagke precedence over those specified here.
```
Comment (cwrw) [inline config/application.rb:40]:
```suggestion
    # Settings in config/environments/* take precedence over those specified here.
```

File: config/environments/test.rb (hunk 6 above)
Comment (cwrw) [reply config/environments/test.rb:57]: ah yes, can remove

File: config/application.rb [hunk 11]
```diff
@@ -5,9 +5,14 @@
...
 require 'active_job/railtie'
 require 'active_record/railtie'
+# require 'active_storage/engine'
 require 'action_controller/railtie'
 require 'action_mailer/railtie'
+# require 'action_mailbox/engine'
+# require 'action_text/engine'
 require 'action_view/railtie'
+# require 'action_cable/engine'
+require 'rails/test_unit/railtie'
```
Comment (cwrw) [inline config/application.rb:15]: we don't need this as we use rspec

File: config/application.rb [hunk 12]
```diff
@@ -83,5 +88,10 @@ class Application < Rails::Application
     config.action_dispatch.default_headers = {
       'Feature-Policy' => "accelerometer 'none'; ambient-light-sensor: 'none', autoplay: 'none', battery: 'none', camera 'none'; display-capture: 'none', document-domain: 'none', fullscreen: 'none', geolocation 'none'; gyroscope 'none'; magnetometer 'none'; microphone 'none'; midi: 'none', payment 'none'; publickey-credentials-get: 'none', usb 'none', wake-lock: 'none', screen-wake-lock: 'none', web-share: 'none'",
     }
+
+    config.active_record.legacy_connection_handling = false
+
+    # todo shouldn't need this unless we add new eager load path
```
Comment (cwrw) [inline config/application.rb:94]: let's remove

File: config/environments/development.rb [hunk 13]
```diff
@@ -55,14 +61,35 @@
   # Print deprecation notices to the Rails logger.
   config.active_support.deprecation = :log
 
+  # todo no need to enable until we populate disallowed deprecations
```
Comment (cwrw) [inline config/environments/development.rb:64]: to remove

File: config/environments/development.rb [hunk 14]
```diff
@@ -55,14 +61,35 @@
...
   config.active_record.verbose_query_logs = true
 
+  # todo not clear what the effect is
+  # Suppress logger output for asset requests.
+  # config.assets.quiet = true
+
+  # todo could use this? doesn't trigger anything when enabled
   # Raises error for missing translations.
-  # config.action_view.raise_on_missing_translations = true
+  config.i18n.raise_on_missing_translations = true
```
Comment (cwrw) [inline config/environments/development.rb:84]: defined in application as per other comment

File: config/environments/development.rb [hunk 15]
```diff
@@ -55,14 +61,35 @@
...
+
+  # todo seems handy, puts template names in the html
+  # Annotate rendered view with file names.
+  config.action_view.annotate_rendered_view_with_filenames = true
+
+  # todo not clear what this does
+  # Uncomment if you wish to allow Action Cable access from any origin.
+  # config.action_cable.disable_request_forgery_protection = true
```
Comment (cwrw) [inline config/environments/development.rb:92]: can remove

File: config/environments/test.rb [hunk 16]
```diff
@@ -39,6 +44,22 @@
   # Print deprecation notices to the stderr.
   config.active_support.deprecation = :stderr
 
+  # todo no need until we add disallowed deprecations
```
Comment (cwrw) [inline config/environments/test.rb:47]: let's remove

File: config/environments/production.rb [hunk 17]
```diff
@@ -36,6 +38,14 @@
     config.action_controller.asset_host = ENV['RAILS_ASSETS_HOST']
   end
 
+  # todo not sure what effect it'll have
+  # Compress CSS using a preprocessor.
+  # config.assets.css_compressor = :sass
+
+  # todo not sure what effect it'll have
+  # Do not fallback to assets pipeline if a precompiled asset is missed.
+  # config.assets.compile = false
```
Comment (cwrw) [inline config/environments/production.rb:47]: let's remove this, and keep to default false

File: config/environments/production.rb [hunk 18]
```diff
@@ -77,6 +87,14 @@
   # Send deprecation notices to registered listeners.
   config.active_support.deprecation = :notify
 
+  # todo seems sensible for prod
+  # Don't log any deprecations.
+  config.active_support.report_deprecations = false
+
+  # todo not sure what effect this will have on current logging
```
Comment (cwrw) [inline config/environments/production.rb:94]: let's remove this as we have a custom semantic logger

**Review by frankieroberto — APPROVED** [review]
Looks good!

Another option for date formatting I discovered recently is stick something like this in the locale file:

```yml
  date:
    formats:
      default: "%-d %B %Y"
      month_and_year: "%B %Y"
```

And then use either `l(date)` for the default formatting or `l(date, format: :month_and_year)` for a non-default one. Saves so many characters!

See [Rails guide: adding date/time formats](https://guides.rubyonrails.org/i18n.html#adding-date-time-formats).

(Why Rails has 2 different ways of doing this, I have no idea...)

**Comment by cwrw** [comment]
Looks great, making a list of things to revisit:
- https://guides.rubyonrails.org/upgrading_ruby_on_rails.html#the-setter-config-autoloader-has-been-deleted
- Confirm autoloading: https://guides.rubyonrails.org/upgrading_ruby_on_rails.html#autoloading-during-initialization
otherwise 👍  after we test on the review app

File: config/environments/production.rb [hunk 19]
```diff
@@ -36,6 +38,14 @@
     config.action_controller.asset_host = ENV['RAILS_ASSETS_HOST']
   end
 
+  # todo not sure what effect it'll have
```
Comment (SamJamCul) [inline config/environments/production.rb:41]: Double check enabling this

File: config/environments/production.rb [hunk 20]
```diff
@@ -151,4 +169,19 @@ def call(env)
...
+  # require "syslog/logger"
+  # config.logger = ActiveSupport::TaggedLogging.new(Syslog::Logger.new "app-name")
+  #
+  # if ENV["RAILS_LOG_TO_STDOUT"].present?
+  #   logger           = ActiveSupport::Logger.new(STDOUT)
+  #   logger.formatter = config.log_formatter
+  #   config.logger    = ActiveSupport::TaggedLogging.new(logger)
+  # end
+
+  # todo probably don't need it in prod
```
Comment (SamJamCul) [inline config/environments/production.rb:164]: Double check

Comment (cwrw) [reply config/environments/production.rb:164]: let's remove

File: config/environments/production.rb (hunk 19 above)
Comment (cwrw) [reply config/environments/production.rb:41]: let's remove

**Review by cwrw — APPROVED** [review]
Tested apply/manage/API flows and all is well 👍

File: app/queries/reasons_for_rejection_count_query.rb (hunk 1 above)
Comment (stevehook) [reply app/queries/reasons_for_rejection_count_query.rb:27]: Looks good to me - that param was removed so redundant.

**Pull request merged**